### PR TITLE
Impressions labels from the Split API

### PR DESCRIPTION
**Pull request by maksimf**

### CODE REVIEW

File: lib/cache/repositories/impressions/memory_repository.rb [hunk 1]
```diff
@@ -18,13 +18,14 @@ def add(split_name, data)
...
-            treatments.each do |split_name, treatment|
+          def add_bulk(key, bucketing_key, treatments_with_labels, time)
+            treatments_with_labels.each do |split_name, treatment_with_label|
               add(
                 split_name,
                 'key_name' => key,
                 'bucketing_key' => bucketing_key,
-                'treatment' => treatment,
+                'treatment' => treatment_with_label[:treatment],
+                'label' => treatment_with_label[:label],
```
Comment (patricioe) [inline lib/cache/repositories/impressions/memory_repository.rb]: what happen to `treatment_with_label[:label]` if label does not exist? Which could be a possibility.

File: lib/engine/parser/split_adapter.rb [hunk 2]
```diff
@@ -56,22 +59,22 @@ def start_based_on_mode(mode)
 
     # Starts thread which loops constantly and stores splits in the splits_repository of choice
     def split_store
-      SplitIoClient::Cache::Stores::SplitStore.new(@splits_repository, @config, @api_key, @metrics, @sdk_blocker).call
+      SplitStore.new(@splits_repository, @config, @api_key, @metrics, @sdk_blocker).call
```
Comment (patricioe) [inline lib/engine/parser/split_adapter.rb]: Thank you ! :)

File: lib/engine/parser/split_treatment.rb [hunk 3]
```diff
@@ -8,29 +8,38 @@ def initialize(segments_repository)
...
 
             next if condition.empty?
 
             if matcher_type(condition).match?(keys[:matching_key], attributes)
-              treatment = Splitter.get_treatment(keys[:bucketing_key], split[:seed], condition.partitions)
+              result = Splitter.get_treatment(keys[:bucketing_key], split[:seed], condition.partitions)
 
-              return treatment.nil? ? default_treatment : treatment
+              if result.nil?
+                return treatment('no rule matched', @default_treatment)
```
Comment (patricioe) [inline lib/engine/parser/split_treatment.rb]: can `'archived'` and `'no rule matched'` as well as `'exception'` and `'killed'` be a constant in one file?

Mostly to follow the pattern or other SDKs.

File: lib/splitclient-rb/clients/split_client.rb [hunk 4]
```diff
@@ -1,114 +1,131 @@
...
-
-    @splits_repository = splits_repository
-    @segments_repository = segments_repository
-    @impressions_repository = impressions_repository
-    @metrics_repository = metrics_repository
-
-    @adapter = adapter
-  end
+    #
+    # Creates a new split client instance that connects to split.io API.
```
Comment (patricioe) [inline lib/splitclient-rb/clients/split_client.rb]: is this just formatting?

File: lib/splitclient-rb/clients/split_client.rb [hunk 5]
```diff
@@ -1,114 +1,131 @@
...
+    # @param split_name [String/Array] name of the feature that is being validated or array of them
+    #
+    # @return [String/Hash] Treatment as String or Hash of treatments in case of array of features
+    def get_treatment(key, split_name, attributes = nil, split_data = nil, store_impressions = true, multiple = false)
+      bucketing_key, matching_key = keys_from_key(key)
+      bucketing_key = matching_key if bucketing_key.nil?
+
+      if matching_key.nil?
+        @config.logger.warn('matching_key was null for split_name: ' + split_name.to_s)
+        return parsed_treatment(multiple, { label: 'matching_key was null', treatment: Treatments::CONTROL })
```
Comment (patricioe) [inline lib/splitclient-rb/clients/split_client.rb]: This is not part of the spec. Make is `exception` for now which is part of the spec..

File: lib/splitclient-rb/clients/split_client.rb [hunk 6]
```diff
@@ -1,114 +1,131 @@
...
-    bucketing_key, matching_key = keys_from_key(key)
-    bucketing_key = matching_key if bucketing_key.nil?
-
-    if matching_key.nil?
-      @config.logger.warn('matching_key was null for split_name: ' + split_name.to_s)
-      return Treatments::CONTROL
-    end
+      if split_name.nil?
+        @config.logger.warn('split_name was null for key: ' + key)
+        return parsed_treatment(multiple, { label: 'split_name was null', treatment: Treatments::CONTROL })
```
Comment (patricioe) [inline lib/splitclient-rb/clients/split_client.rb]: when the split_name is nil, there is not impression right? what happens in line 57 then?

` 'split_name was null'` is not part of the spec.

File: lib/splitclient-rb/clients/split_client.rb [hunk 7]
```diff
@@ -1,114 +1,131 @@
...
 
-    treatments
-  end
+    #
+    # obtains the treatment for a given feature
+    #
+    # @param key [String/Hash] user id or hash with matching_key/bucketing_key
+    # @param split_name [String/Array] name of the feature that is being validated or array of them
+    #
+    # @return [String/Hash] Treatment as String or Hash of treatments in case of array of features
```
Comment (patricioe) [inline lib/splitclient-rb/clients/split_client.rb]: we are missing params in this ruby doc.

File: spec/engine/parser/split_treatment_spec.rb [hunk 8]
```diff
@@ -0,0 +1,21 @@
+require 'spec_helper'
+
+include SplitIoClient
+
+describe Engine::Parser::SplitTreatment do
```
Comment (patricioe) [inline spec/engine/parser/split_treatment_spec.rb]: can you add one more for `exception` ?

File: lib/cache/repositories/impressions/memory_repository.rb (hunk 1 above)
Comment (maksimf) [inline lib/cache/repositories/impressions/memory_repository.rb]: It will be `nil`

File: lib/engine/parser/split_adapter.rb (hunk 2 above)
Comment (maksimf) [inline lib/engine/parser/split_adapter.rb]: Welcome :-)

File: spec/engine/parser/split_treatment_spec.rb (hunk 8 above)
Comment (maksimf) [inline spec/engine/parser/split_treatment_spec.rb]: I've added those specs to the `engine_spec`, lines 31-70

**Comment by maksimf** [comment]
@patricioe All fixed

**Comment by maksimf** [comment]
@patricioe This PR looks fine to me, please check it out

**Pull request merged**